### PR TITLE
Dockerize Laravel Stack

### LINKED ISSUE #105: Dockerize the Laravel / PHP / Dev Stack
@MarkMcDaniels brought up the challenges of jumping into the Code For Greenville projects due to the dev setup tasks and effort.

During the last [CFG meeting](https://codeforgreenville.org/) @jeremywight @allella and @cscully thought we should Dockerize some of the projects. 

This HG repo may be a good first candidate.

We talked a bit about it already on the [Slack](https://app.slack.com/client/TDPR0G67P/CDP1BRXKK?cdn_fallback=1), but I figured to leave this open for some planning and in case others are interested.

### PR DESCRIPTION
**Pull request by oliviasculley**
This PR attempts to dockerize the Laravel stack, and provides some documentation to hopefully initialize a reproducible docker setup from scratch.

Please let me know if you run into any issues trying to initialize the project through docker, or if you have any other comments. Thanks!

Closes #105

### CODE REVIEW

**Comment by zach2825** [comment]
Looks good, just a little feedback.

I resolved a conflict coming from the develop branch.

Check out [Laravel sail](https://laravel.com/docs/8.x/sail#starting-and-stopping-sail). laravel sail has a bunch of built-in helpers to use docker. Like `./vendor/bin/sail up` starts the docker, and `./vendor/bin/sail down` stop them. There are other useful commands.

You can also do things like the artisan commands, but with sail like `./vendor/bin/sail artisan migrate` for example
https://laravel.com/docs/8.x/sail#executing-sail-commands

**Comment by oliviasculley** [comment]
Sail is built from docker pretty extensively, and is what I referenced heavily for this PR. https://tech.osteel.me/posts/you-dont-need-laravel-sail was super helpful to me in understanding what Sail was, but I think just using a stock docker setup uncovers a bit less of the magic, is more accessible, and will let people iterate and debug easier

**Comment by zach2825** [comment]
That's fair, to be honest I've never had luck using docker. I like what you
did here it loads up and servers the app just if you need to do artisan
commands sail makes that easier.

Thank you 😊

**Review by zach2825 — APPROVED** [review]
Docker loads for me, I'm able to navigate the site in it. I'm not sure if anything else is needed but it's working

**Review by allella — COMMENTED** [review]
My only question is if we need the .env.docker.  Is that a common pattern? If not, could we move the Docker specific configuration into .env.example and suggest people uncomment and/or adjust Docker config lines in there .env before running Docker.

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -137,6 +137,24 @@ The HackGreenville project is now up and running! You should be able to open [lo
 The `composer install` command will run `php artisan migrate --seed; yarn install; yarn prod` which will build the project. 
 To develop you'll want to run `php artisan serve` to start the applications php server and in another terminal you'll want to run `yarn watch` to watch for frontend resource changes and re-build them when detected. 
 
+#### Initial Setup (Docker)
+The docker setup of this project should only be done for advanced users, or if
+needed for runtime compatibility issues.
+
+You need to make a copy of the `.env.docker` file and rename it to `.env` at the
```
[the remark this answers is not in the document]
Comment (csculley) [reply CONTRIBUTING.md:144]: I think I've included some env var changes to `.env.docker` that are docker specific, so I put them in `.env.docker` as not to mess with normal host installs. It's a common pattern from what I've seen, where people copy either `.env.example` or `.env.docker` to `.env` which is usually gitignored and contains secrets

[the remark this answers is not in the document]
Comment (oliviasculley) [reply CONTRIBUTING.md:144]: I think I've included some env var changes to `.env.docker` that are docker specific, so I put them in `.env.docker` as not to mess with normal host installs. It's a common pattern from what I've seen, where people copy either `.env.example` or `.env.docker` to `.env` which is usually gitignored and contains secrets

**Comment by allella** [comment]
Alright, I'm giving it a try and getting an error on _docker-compose pull_
Is this a temporary network error?

image

**Comment by zach2825** [comment]
I got that too, one time all 3 failed. re-try a few more times to see if it works for you

**Comment by allella** [comment]
Still not working for me.

The docker-compose.yml has a [context](https://docs.docker.com/compose/compose-file/compose-file-v3/#context) of

`context: ./vendor/laravel/sail/runtimes/8.0`

but that directory doesn't exist if I clone the repo and continue with the Docker steps.

Does composer need to be run locally in order for the Sail image to be referenced in this way?

It seems Docker is trying to pull an image from a local directory and since it doesn't exist I'm guessing it's treating the context as a URL, which is likely why I'm getting the error

```
ERROR: for laravel.test  Get https://sail-8.0/v2/: dial tcp: lookup sail-8.0: no such host
ERROR: Get https://sail-8.0/v2/: dial tcp: lookup sail-8.0: no such host

```

> Either a path to a directory containing a Dockerfile, or a url to a git repository.
> 
> When the value supplied is a relative path, it is interpreted as relative to the location of the Compose file. This directory is also the build context that is sent to the Docker daemon.

**Comment by oliviasculley** [comment]
I forgot to include instructions on how to fetch the Sail dependencies, so I've added instructions that utilize composer and alternatively just plain git to fetch the needed dependencies before building. I've also added additional instructions regarding the file permissions and build flag

**Comment by allella** [comment]
My issues have been resolved after @oliviasculley 's latest commits.

**Comment by allella** [comment]
Just noticed this was to the master branch.  We have been using the develop branch as the main branch for the stage website. Probably good for one of us to merge master back to develop to sync up. Or, perhaps there's a better branching strategy.

**Comment by oliviasculley** [comment]
I can also open a PR to `develop` if you'd like!

**Comment by allella** [comment]
Sure. We've been merging PRs into develop and then I'd do a release merge from develop to master after testing was done on the staging website.

**Pull request merged**